modal: Add options to selectively mark messages as read.

AMANVISHWAKARMA27 opened the pull request:
Added a modal with a dropdown to selectively mark messages as read. The modal includes three dropdown options:

> - [ ] Muted topics
> - [x] Topics I don't follow (default checked)
> - [ ] All messages

Below the select options, there is a line that displays the count of messages that will be marked as read:

> N message/messages will be marked as read. (This updates dynamically when the selection changes.)

</br>

## **Difference between the previous and the present model**

<table>
  <tr>
    <th>Before</th>
    <th>After</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/75223fd5-afbc-4227-bded-a619730088df" width="300"></td>
    <td><img src="https://github.com/user-attachments/assets/7a5d1810-36f9-48fb-b5c1-7c3a4883cef8" width="300"></td>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/9078691c-09ae-46d1-bd9b-30748c4ae431" width="300"></td>
    <td><img src="https://github.com/user-attachments/assets/4f6b2400-e4a1-442e-8a61-e688d0a8d46f" width="300"></td>
  </tr>
<tr>
<td><img src="https://github.com/user-attachments/assets/d2ba987e-7ebe-4fd6-ba87-9deacb3e1dfa" width="300"></td>
<td><img src="https://github.com/user-attachments/assets/74ac2047-3f24-4cd0-a368-5cbc7458d4a8" width="300"></td>
</tr>
<tr>
<td><img src="https://github.com/user-attachments/assets/dd9d0dcc-d452-44d5-957d-401258c99e7c" width="300"></td>
<td><img src="https://github.com/user-attachments/assets/58b0d925-692c-4179-b505-27047f752569" width="300"></td>
</tr>
</table>

</br>

## **Demonstrating how the count changes on change of select option**

<table>
  <tr>
    <th>0 messages</th>
    <th>1 message</th>
    <th>More than 1 message</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/b555f8f3-5e54-4921-9aa9-445d032189af" width="250"></td>
    <td><img src="https://github.com/user-attachments/assets/3c9ef673-76cf-4719-bcfe-15d85cf8b07f" width="250"></td>
    <td><img src="https://github.com/user-attachments/assets/e01b73e4-94cb-4662-af86-2d5803e49d61" width="250"></td>
  </tr>
</table>

<br/>

## **Tasks completed**

- [x] Rename "Mark all messages as read" -> "Mark messages as read" in the menu.
- [x] Change the confirmation modal.

Fixes : #30025 

<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
I'm pretty sure this is supposed to be a dropdown with those options, not 3 checkboxes.

AMANVISHWAKARMA27 commented:
> I'm pretty sure this is supposed to be a dropdown with those options, not 3 checkboxes.

@timabbott , @alya has clearly stated in the [Issue](#30025 ) that a modal is required.Kindly check it.


![image](https://github.com/user-attachments/assets/0859553f-1045-47d0-be69-c072bc16f96e)

alya commented:
Yes, as @timabbott said, it should be a modal with a dropdown.

AMANVISHWAKARMA27 commented:
> Yes, as @timabbott said, it should be a modal with a dropdown.

Thanks for the clarification @alya . 
One more confirmation, should it be multi-select or mono-select??

AMANVISHWAKARMA27 commented:
Hey @alya, @timabbott , as per your instructions I have changed the checkbox to a dropdown including a statement below the select menu to show the message count. Please check it out.

<details>
<summary> Screenshot </summary>

![image](https://github.com/user-attachments/assets/1d8465e8-d0ea-4423-976c-3d8d54745a20)


</details>

<details>

<summary> Video demonstration </summary>

https://github.com/user-attachments/assets/f63939ee-d120-4c5f-9e2f-1b1370bbcdf7

</details>

AMANVISHWAKARMA27 commented:
@alya, please review the changes.

alya commented:
The styling for the dropdown looks strange. It should look like other dropdowns in modals, e.g., in the **Invite users to organization** modal.

alya commented:
Also, be careful to avoid typos, please!

AMANVISHWAKARMA27 commented:
> The styling for the dropdown looks strange. It should look like other dropdowns in modals, e.g., in the **Invite users to organization** modal.

Hey @alya , I have made the changes as per your suggestions. Kindly review them.

![image](https://github.com/user-attachments/assets/92b78109-c64b-4109-b466-05ed5d744c9b)

![Screenshot from 2024-10-04 11-47-04](https://github.com/user-attachments/assets/37b765db-44c3-414e-9e5c-e81cb14e9f18)

alya commented:
Did you check your work for typos, as I suggested? `I` is always capitalized.

AMANVISHWAKARMA27 commented:
> Did you check your work for typos, as I suggested? `I` is always capitalized.

@alya, I know but it failed the capitalization test when being captial so made it small and hence passed.
Still, I will again make it capital and push it again.

AMANVISHWAKARMA27 commented:
Hey @alya , when I capitalized "i" in "Topics i dont follow", then it failed the capitalization test. Pleae check.

![image](https://github.com/user-attachments/assets/7bf9b5e6-4796-4dcb-bb90-96489acf7729)

alya commented:
The requirement is to user proper spelling in the UI. If you can't figure out how to make tests, pass, you can post a specific question in the #frontend stream in the [Zulip development community](https://zulip.com/development-community/). Be sure to follow https://zulip.readthedocs.io/en/latest/contributing/asking-great-questions.html. Thanks!

AMANVISHWAKARMA27 commented:
Thanks for the suggestion but I posted this issue in `#Issue` stream and got the solution.
[CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/Capitalization.20Issue) Thread

I just had to add "Topics I don't follow" in the file `tools/lib/capitalization.py` .
Now the PR is ready for another review.

AMANVISHWAKARMA27 commented:
Hey @alya , kindly review the PR.

alya commented:
Thanks! The text in the menu and the modal needs to follow the spec in the issue. It wastes reviewer time to ask for feedback when you have not made sure that this is the case, so please be mindful to check your own work if you'd like us to be able to review your PRs.

alya commented:
You will also need to handle pluralization properly, making sure it's correctly tagged for translation. You can see the read receipts modal for an example of how this is done.
![Screenshot 2024-10-07 at 12 39 32@2x](https://github.com/user-attachments/assets/757070e9-7f2b-44ff-a5f3-78b0b4113dc9)

AMANVISHWAKARMA27 commented:
I got it @alya . Just one more thing, can I write it as message(s) kind something. I have seen this on many websites to keep it correct and reduce complexity.

AMANVISHWAKARMA27 commented:
Else I have plan, that the word "message or messages" can be conditionally rendered as a prop. Its not that hard to implement although.

AMANVISHWAKARMA27 commented:
> You will also need to handle pluralization properly, making sure it's correctly tagged for translation. You can see the read receipts modal for an example of how this is done. ![Screenshot 2024-10-07 at 12 39 32@2x](https://private-user-images.githubusercontent.com/2090066/374307655-757070e9-7f2b-44ff-a5f3-78b0b4113dc9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgzMzQ2MzgsIm5iZiI6MTcyODMzNDMzOCwicGF0aCI6Ii8yMDkwMDY2LzM3NDMwNzY1NS03NTcwNzBlOS03ZjJiLTQ0ZmYtYTVmMy03OGIwYjQxMTNkYzkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMDdUMjA1MjE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmYxMWZkODZkOTJhNjMyNDJiYjYxZmZmZTM5MmRlMmI5YWViZDg1Mzc2M2RlZjlhMzc5ZDMxNzU4MTM4YTA5NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Ec-tP4IhEQJtAqF_LltlAaMUAUQYCLbCtuGD-QNb-EI)

Hey @alya, now the message dialog box handles pluralization properly. you may check it.

![image](https://github.com/user-attachments/assets/70aeda41-c360-4bb0-8553-371bbdd579c6)

AMANVISHWAKARMA27 commented:
Hey @alya kindly review the changes.

alya commented:
Your implementation will not work correctly in translation. As I noted above:

> You will also need to handle pluralization properly, making sure it's correctly tagged for translation. You can see the read receipts modal for an example of how this is done.

AMANVISHWAKARMA27 commented:
> Your implementation will not work correctly in translation. As I noted above:
> 
> > You will also need to handle pluralization properly, making sure it's correctly tagged for translation. You can see the read receipts modal for an example of how this is done.

Got it. Will make the changes soon.
Can you provide me where can I find this - "read receipts modal ".

alya commented:
https://zulip.com/help/read-receipts

You can `git grep` to find the code.

AMANVISHWAKARMA27 commented:
Hey @alya, as per your suggestions I have made the changes. You may review it.

For the purpose of pluralization and translation I have made the use of i18n and i18n.t .

Let me know if more changes are required.

AMANVISHWAKARMA27 commented:
Hey @alya, please review the changes.

AMANVISHWAKARMA27 commented:
hey @alya, please review the latest changes in the PR.

alya commented:
You don't need to post screenshots of code -- that's never a good way to review it.

alya commented:
You need to put updated screenshots in the PR description, making sure the text exactly follows the spec in the issue. Include examples with 0, 1 and more messages.

AMANVISHWAKARMA27 commented:
Hey @alya , I have updated the PR description as per your suggestions. Kindly have a look .

AMANVISHWAKARMA27 commented:
@alya, you told that when count is 1, 'message' should be there and for count more than 1, 'messages'. 
I wonder what about the case when count is 0. Should it be 'message' or 'messages'?

AMANVISHWAKARMA27 commented:
Also, I have made the changes in the PR description as per your suggestions. You may have a look.

AMANVISHWAKARMA27 commented:
@alya , please review the changes.

AMANVISHWAKARMA27 commented:
Hey @alya. can you please review the PR.

alya commented:
Thanks, but your commit message still does not match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages). You also have other typos in the commit message, e.g., "Fixes: #30025' "

If you would like us to be able to review your work going forward, you need to be careful to review your own work thoroughly before requesting feedback.

AMANVISHWAKARMA27 commented:
Thank you for your feedback, @alya. I have revised and enhanced the commit message according to your suggestions. Please review it, and I hope this version meets your standards for the PR.

AMANVISHWAKARMA27 commented:
@alya , can I get some comments on the solution of the issue?

amanagr commented on web/src/unread_ops.ts:
```diff
@@ -52,13 +56,190 @@ export function confirm_mark_all_as_read(): void {
+    }
+
+    if (selected_option === "muted_topics") {
+        is_processing = true;
+        const all_streams = stream_data.subscribed_stream_ids();
+
+        for (const stream_id of all_streams) {
+            const topics = stream_topic_history.get_recent_topic_names(stream_id);
+            const stream_name = stream_data.get_stream_name_from_id(stream_id);
+            for (const topic of topics) {
+                const is_muted = get_topic_visibility_policy(stream_id, topic) === 1;
+                if (is_muted) {
+                    bulk_update_read_flags_for_narrow(
+                        [
+                            {operator: "stream", operand: stream_name},
+                            {operator: "topic", operand: topic},
+                        ],
+                        "add",
+                        {},
+                        modal_id,
+                    );
+                }
+            }
+        }
+        processing_done = true;
+    }
+
+    if (selected_option === "topics_not_followed") {
+        is_processing = true;
+        const narrow_options: NarrowTerm[] = [{operator: "is", operand: "unread", negated: false}];
+        narrow_options.push({operator: "is", operand: "followed", negated: true});
+        mark_as_read_with_narrow(narrow_options, modal_id);
+        processing_done = true;
```
Please read `unread` library and get the unread messages which are relevant as per the selected option and ask server to mark them as read.

AMANVISHWAKARMA27 replied:
I am not able to get the unread topics from streams.

AMANVISHWAKARMA27 replied:
I have optimized the implementation.

AMANVISHWAKARMA27 commented on web/src/unread_ops.ts:
```diff
@@ -52,13 +56,190 @@ export function confirm_mark_all_as_read(): void {
         html_body,
         on_click() {
-            mark_all_as_read(modal_id);
+            mark_selected_options_as_read(modal_id);
```
Changed `mark_selected_option_as_read` to `mark_messages_as_read`. 
Kindly suggest if there can be a better name.

AMANVISHWAKARMA27 commented on web/src/unread_ops.ts:
```diff
@@ -52,13 +56,190 @@ export function confirm_mark_all_as_read(): void {
         loading_spinner: true,
     });
+
+    const default_message_count = countMessages("topics_not_followed");
```
Changed `snakeCase` to `snake_case`.

AMANVISHWAKARMA27 commented on web/src/unread_ops.ts:
```diff
@@ -52,13 +56,190 @@ export function confirm_mark_all_as_read(): void {
+    );
+
+    $("#mark_as_read_option").on("change", function () {
+        const selected_option = String($(this).val());
```
The function `countMessages()` accepts string, bu the value I am getting from `$(this).val` can be of type `string`, `number`, `string[]` or `undefined`. So, I converted every value to string before passing it to the function. 

Although I am trying to implement it in a different way so that type conversion will not be required.

AMANVISHWAKARMA27 commented on web/src/unread_ops.ts:
```diff
@@ -52,13 +56,190 @@ export function confirm_mark_all_as_read(): void {
+
+    switch (selected_option) {
+        case "all_messages": {
+            count = getAllMessagesCount();
```
Yes @amanagr , I have also done the same thing. But as you pointed it out, I have wrote redundant code, as I could directly call the function instead of defining another function that call the `unread.get_unread_message_count()`.

Corrected!

AMANVISHWAKARMA27 commented on web/src/unread_ops.ts:
```diff
@@ -52,13 +56,190 @@ export function confirm_mark_all_as_read(): void {
+}
+
+function getMutedTopicMessagesCount(): number {
+    let count = 0;
```
Alright, I will try to write a more optimized method.

AMANVISHWAKARMA27 replied:
Corrected !

AMANVISHWAKARMA27 commented on web/src/unread_ops.ts:
```diff
@@ -52,13 +56,190 @@ export function confirm_mark_all_as_read(): void {
+}
+
+function getUnfollowedTopicMessagesCount(): number {
+    let count = 0;
```
Implementation : 

**For all unread message count** :  `unread.get_unread_message_count()`
**For all unread messages from muted topics**:  `unread.get_counts().home_unread_messages`
**For all unread messages from unfollowed topics**: `unread.get_unread_message_count() - unread.get_counts().followed_topic_unread_messages_count`

AMANVISHWAKARMA27 replied:
Corrected !

AMANVISHWAKARMA27 commented on web/src/unread_ops.ts:
```diff
@@ -52,13 +56,190 @@ export function confirm_mark_all_as_read(): void {
+    }
+
+    if (selected_option === "all_messages") {
+        is_processing = true;
+        mark_all_as_read(modal_id);
+        processing_done = true;
```
I have fixed this. So, I removed the extra loading banner added by me and just used `loading_spinner: true` . Since, this is already implemented in the `bulk_update_read_flags_for_narrow` , this is implemented easily.

AMANVISHWAKARMA27 commented:
Hey @amanagr , I have updated the implementation as per your suggestions. The PR is ready for another review.

AMANVISHWAKARMA27 commented:
Hey @amanagr , I have updated the implementation as per your suggestions. The PR is ready for another review.

AMANVISHWAKARMA27 commented:
@amanagr, in the mean time I made few more changes:

1. **Variable and function names**: I changed the names of the functions created by me to more appropriate and self-explaining names.
2.  **File naming**: Initially the handlebar file responsible for the dialog box had name - `confirm_mark_all_as_read.hbs`, I changes its name into the appropriate name according to the current scenario - `confirm_mark_messages_as_read.hbs`.
3.  **When muted topics are 0**: When there are no  unread messages in muted topics, clicking on muted topics option does not stop the loader and hide the dialog box. Now this is fixed.
4. **DM Messages in unfollowd topic**: On further testing, I found that the logic counts the DMs in unfollowed topics. This was not my fault as the narrow option I used perform in this manner. Now I added another narrow option `-is:dm` to filter out the DMs.
5.  **Unfollowed topics message count logic**: Subtracted the count of direct messages from the count of unfollowed topics messages to get the correct count.

AMANVISHWAKARMA27 commented:
In the meantime, I’ve made several changes, including the ones you suggested, along with a few additional improvements that I found necessary. The PR is now ready for another review. Thanks!

AMANVISHWAKARMA27 commented:
@amanagr I have made some changes by looking at the codebase again and closed PRs.It would be really helpful if u could review the PR again.

alya commented:
I'm closing this PR, as it's not on track to being something we can merge. Thanks for giving this a try, @AMANVISHWAKARMA27 . I recommend getting more solid on your coding skills first, so that you are able to contribute more effectively to open source projects.

whilstsomebody commented:
Hey @AMANVISHWAKARMA27 , can I take up this issue and continue with your PR if you don' t mind?

AMANVISHWAKARMA27 commented:
> Hey @AMANVISHWAKARMA27 , can I take up this issue and continue with your PR if you don' t mind?

Sure, please don't forget to read the contribution guide before proceeding.
